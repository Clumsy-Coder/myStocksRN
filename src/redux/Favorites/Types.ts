import { Action } from 'redux';

/*
  ██████╗ ███████╗██████╗ ██╗   ██╗ ██████╗███████╗██████╗     ███████╗████████╗ █████╗ ████████╗███████╗
  ██╔══██╗██╔════╝██╔══██╗██║   ██║██╔════╝██╔════╝██╔══██╗    ██╔════╝╚══██╔══╝██╔══██╗╚══██╔══╝██╔════╝
  ██████╔╝█████╗  ██║  ██║██║   ██║██║     █████╗  ██████╔╝    ███████╗   ██║   ███████║   ██║   █████╗
  ██╔══██╗██╔══╝  ██║  ██║██║   ██║██║     ██╔══╝  ██╔══██╗    ╚════██║   ██║   ██╔══██║   ██║   ██╔══╝
  ██║  ██║███████╗██████╔╝╚██████╔╝╚██████╗███████╗██║  ██║    ███████║   ██║   ██║  ██║   ██║   ███████╗
  ╚═╝  ╚═╝╚══════╝╚═════╝  ╚═════╝  ╚═════╝╚══════╝╚═╝  ╚═╝    ╚══════╝   ╚═╝   ╚═╝  ╚═╝   ╚═╝   ╚══════╝

*/

/**
 * Favorites reducer state
 */
export interface FavoritesReducerState {
  symbols: string[];
}

// ---------------------------------------------------------------------------------------------- //
/*
   █████╗  ██████╗████████╗██╗ ██████╗ ███╗   ██╗    ████████╗██╗   ██╗██████╗ ███████╗███████╗
  ██╔══██╗██╔════╝╚══██╔══╝██║██╔═══██╗████╗  ██║    ╚══██╔══╝╚██╗ ██╔╝██╔══██╗██╔════╝██╔════╝
  ███████║██║        ██║   ██║██║   ██║██╔██╗ ██║       ██║    ╚████╔╝ ██████╔╝█████╗  ███████╗
  ██╔══██║██║        ██║   ██║██║   ██║██║╚██╗██║       ██║     ╚██╔╝  ██╔═══╝ ██╔══╝  ╚════██║
  ██║  ██║╚██████╗   ██║   ██║╚██████╔╝██║ ╚████║       ██║      ██║   ██║     ███████╗███████║
  ╚═╝  ╚═╝ ╚═════╝   ╚═╝   ╚═╝ ╚═════╝ ╚═╝  ╚═══╝       ╚═╝      ╚═╝   ╚═╝     ╚══════╝╚══════╝

*/

/**
 * Favorites action types
 */
export enum ActionTypes {
  ADD_FAVORITE_STOCK = 'FAVORITES/ADD_FAVORITE_STOCK',
  REMOVE_FAVORITE_STOCK = 'FAVORITES/REMOVE_FAVORITE_STOCK',
  CLEAR_FAVORITE_STOCKS = 'FAVORITES/CLEAR_FAVORITE_STOCKS',
}

// ---------------------------------------------------------------------------------------------- //
/*
   █████╗  ██████╗████████╗██╗ ██████╗ ███╗   ██╗███████╗
  ██╔══██╗██╔════╝╚══██╔══╝██║██╔═══██╗████╗  ██║██╔════╝
  ███████║██║        ██║   ██║██║   ██║██╔██╗ ██║███████╗
  ██╔══██║██║        ██║   ██║██║   ██║██║╚██╗██║╚════██║
  ██║  ██║╚██████╗   ██║   ██║╚██████╔╝██║ ╚████║███████║
  ╚═╝  ╚═╝ ╚═════╝   ╚═╝   ╚═╝ ╚═════╝ ╚═╝  ╚═══╝╚══════╝

*/

/**
 * Add stock to Favorites action
 */
export interface AddFavoriteStockAction extends Action<ActionTypes.ADD_FAVORITE_STOCK> {
  type: ActionTypes.ADD_FAVORITE_STOCK;
  readonly stockSymbol: string;
}

/**
 * Remove stock from Favorites action
 */
export interface RemoveFavoriteStockAction extends Action<ActionTypes.REMOVE_FAVORITE_STOCK> {
  type: ActionTypes.REMOVE_FAVORITE_STOCK;
  readonly stockSymbol: string;
}

/**
 * Clear all stocks from Favorites action
 */
export interface ClearFavoriteStocksAction extends Action<ActionTypes.CLEAR_FAVORITE_STOCKS> {
  type: ActionTypes.CLEAR_FAVORITE_STOCKS;
}

/**
 * Union of actions for AddFavoriteStock, RemoveFavoriteStock, ClearFavoriteStocks
 */
export type FavoritesActions =
  | AddFavoriteStockAction
  | RemoveFavoriteStockAction
  | ClearFavoriteStocksAction;
