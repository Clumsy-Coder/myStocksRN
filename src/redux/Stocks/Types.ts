/**
 * Data format when fetching stock quote
 * check https://iexcloud.io/docs/api/#quote
 *
 * @interface IStockQuote
 */
export interface IStockQuote {
  symbol: string; // stock ticker. ex: AAPL
  companyName: string; // company name
  primaryExchange: string; // Refers to the primary listing exchange for the symbol
  latestPrice: number; // Refers to the latest relevant price of the security which is derived from multiple sources
  latestVolume: number; // Refers to the latest total market volume of the stock across all markets
  latestUpdate: number; // latest update time in epoch represented in milliseconds
  latestTime: string; // latest update time in human readable format
  change: number; // Refers to the change in price between latestPrice and previousClose
  changePercent: number; // Refers to the percent change in price between latestPrice and previousClose
  open: number; // open stock price
  close: number; // close stock price
  high: number | null; // highest stock price during stock hours. null before 9:45 and weekends
  low: number | null; // lowest stock price during stock hours. null before 9:45 and weekends
  week52High: number; // highest stock price in the last 52 weeks
  week52Low: number; // lowest stock price in the last 52 weeks
  marketCap: number; // is calculated in real time using latestPrice
  ytdChange: number; // Refers to the price change percentage from start of year to previous close
  peRatio: number; // Refers to the price-to-earnings ratio for the company
  // volume: number; // Total volume for the stock, but only updated after market open
  avgTotalVolume: number;
  previousClose: number; // Refers to the previous trading day closing price
  extendedPrice: number; // Refers to the 15 minute delayed price outside normal market hours 0400 - 0930 ET and 1600 - 2000 ET
  extendedChange: number; // Refers to the price change between extendedPrice and latestPrice
  extendedChangePercent: number; // Refers to the price change percent between extendedPrice and latestPrice
  previousVolume: number; // Refers to the previous trading day volume

  // openTime: number;
  // closeTime: number;
  // openSource: string;
  // closeSource: string;
  // highTime: number;
  // highSource: string;
  // lowTime: number;
  // lowSource: string;

  // delayedPrice: number;
  // delayedPriceTime: number;
  // oddLotDelayedPrice: number;
  // oddLotDelayedPriceTime: number;
  // extendedPriceTime: number;
  // iexRealtimePrice: number;
  // iexRealtimeSize: number;
  // iexLastUpdated: number;
  // iexMarketPercent: number;
  // iexVolume: number;
  // iexBidPrice: number;
  // iexBidSize: number;
  // iexAskPrice: number;
  // iexAskSize: number;
  // iexOpen: null;
  // iexOpenTime: null;
  // iexClose: number;
  // iexCloseTime: number;
  // lastTradeTime: number;
  // isUSMarketOpen: Boolean;
}

/**
 * Data format when fetching stock chart data.
 * check https://iexcloud.io/docs/api/#charts
 * check https://iexcloud.io/docs/api/#historical-prices
 * check https://iexcloud.io/docs/api/#intraday-prices
 *
 * @interface IStockChart
 */
export interface IStockChart {
  date: string; // Formatted as YYYY-MM-DD
  open: number; // Adjusted data for historical dates. Split adjusted only
  close: number; // Adjusted data for historical dates. Split adjusted only
  high: number; // Adjusted data for historical dates. Split adjusted only
  low: number; // Adjusted data for historical dates. Split adjusted only
  volume: number; // Adjusted data for historical dates. Split adjusted only
  currency: string;
  change: number; // Change from previous trading day
  changePercent: number; // Change percent from previous trading day
  label: string; // A human readable format of the date depending on the range
  changeOverTime: number; // Percent change of each interval relative to first value. Useful for comparing multiple stocks
  // uClose: number; // Unadjusted data for historical dates
  // uOpen: number; // Unadjusted data for historical dates
  // uHigh: number; // Unadjusted data for historical dates
  // uLow: number; // Unadjusted data for historical dates
  // uVolume: number; // Unadjusted data for historical dates
}

// ---------------------------------------------------------------------------------------------- //
// used for Stocks reducer
export interface IStockState {
  [symbol: string]: {
    quote: {
      fetching: Boolean;
      data: IStockQuote | null;
      error?: any;
    };
    chart?: {
      fetching: boolean;
      data: IStockChart[] | null;
      error?: any;
    };
  };
}

// ---------------------------------------------------------------------------------------------- //
// action types

export enum ActionTypesEnum {
  GET_STOCK_QUOTE = 'STOCKS/GET_STOCK_QUOTE',
  GET_STOCK_QUOTE_PENDING = 'STOCKS/GET_STOCK_QUOTE_PENDING',
  GET_STOCK_QUOTE_FULFILLED = 'STOCKS/GET_STOCK_QUOTE_FULFILLED',
  GET_STOCK_QUOTE_REJECTED = 'STOCKS/GET_STOCK_QUOTE_REJECTED',

  GET_STOCK_CHART = 'STOCKS/GET_STOCK_CHART',
  GET_STOCK_CHART_PENDING = 'STOCKS/GET_STOCK_CHART_PENDING',
  GET_STOCK_CHART_FULFILLED = 'STOCKS/GET_STOCK_CHART_FULFILLED',
  GET_STOCK_CHART_REJECTED = 'STOCKS/GET_STOCK_CHART_REJECTED',
}

// ---------------------------------------------------------------------------------------------- //
// action creators

// get stock quote action creator
// called to initiate the fetching process
export interface IGetStockQuoteAC {
  type: ActionTypesEnum.GET_STOCK_QUOTE;
  readonly stockSymbol: string; // stock symbol
}

// called only in redux saga to handle side effects
export interface IGetStockQuotePendingAC {
  type: ActionTypesEnum.GET_STOCK_QUOTE_PENDING;
  readonly stockSymbol: string;
}

// called only in redux saga to handle side effects
export interface IGetStockQuoteFulfilledAC {
  type: ActionTypesEnum.GET_STOCK_QUOTE_FULFILLED;
  readonly stockSymbol: string;
  readonly payload: { data: IStockQuote };
}

// called only in redux saga to handle side effects
export interface IGetStockQuoteRejectedAC {
  type: ActionTypesEnum.GET_STOCK_QUOTE_REJECTED;
  readonly stockSymbol: string;
  readonly error: any;
}

export type StockQuoteAction =
  | IGetStockQuoteAC
  | IGetStockQuotePendingAC
  | IGetStockQuoteFulfilledAC
  | IGetStockQuoteRejectedAC;

// ---------------------------------------------------------------------------------------------- //
// get stock chart action creator
// called to initiate the fetching process
export interface IGetStockChartAC {
  type: ActionTypesEnum.GET_STOCK_CHART;
  readonly stockSymbol: string;
}

// called only in redux saga to handle side effects
export interface IGetStockChartPendingAC {
  type: ActionTypesEnum.GET_STOCK_CHART_PENDING;
  readonly stockSymbol: string;
}

// called only in redux saga to handle side effects
export interface IGetStockChartFulfilledAC {
  type: ActionTypesEnum.GET_STOCK_CHART_FULFILLED;
  readonly stockSymbol: string;
  readonly payload: { data: IStockChart[] };
}

// called only in redux saga to handle side effects
export interface IGetStockChartRejectedAC {
  type: ActionTypesEnum.GET_STOCK_CHART_REJECTED;
  readonly stockSymbol: string;
  readonly error: any;
}

export type StockChartAction =
  | IGetStockChartAC
  | IGetStockChartPendingAC
  | IGetStockChartFulfilledAC
  | IGetStockChartRejectedAC;

// ---------------------------------------------------------------------------------------------- //
// all Stock action creators
export type StockActionTypes = StockQuoteAction | StockChartAction;
