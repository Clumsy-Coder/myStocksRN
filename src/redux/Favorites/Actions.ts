import { ActionCreator } from 'redux';

import {
  ActionTypes,
  AddFavoriteStockAction,
  RemoveFavoriteStockAction,
  ClearFavoriteStocksAction,
} from 'src/redux/Favorites/Types';

/**
 * Add stock to Favorites action creator.
 * @param stockSymbol - Stock symbol in uppercase. Ex: AAPL
 */
export const addFavoriteStock: ActionCreator<AddFavoriteStockAction> = (
  stockSymbol: string,
): AddFavoriteStockAction => ({
  type: ActionTypes.ADD_FAVORITE_STOCK,
  stockSymbol,
});

/**
 * Remove stock from Favorites action creator.
 * @param stockSymbol - Stock symbol in uppercase. Ex: AAPL
 */
export const removeFavoriteStock: ActionCreator<RemoveFavoriteStockAction> = (
  stockSymbol: string,
): RemoveFavoriteStockAction => ({
  type: ActionTypes.REMOVE_FAVORITE_STOCK,
  stockSymbol,
});

/**
 * Clear all stocks from Favorites action creator.
 */
// eslint-disable-next-line max-len
export const clearFavoriteStocks: ActionCreator<ClearFavoriteStocksAction> = (): ClearFavoriteStocksAction => ({
  type: ActionTypes.CLEAR_FAVORITE_STOCKS,
});
