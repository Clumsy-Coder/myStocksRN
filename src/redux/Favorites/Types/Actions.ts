import { Action } from 'redux';

import ActionTypes from '@redux/Favorites/Types/ActionTypes';

/**
 * Add stock to Favorites action
 */
export interface AddFavoriteStockAction extends Action<ActionTypes.ADD_FAVORITE_STOCK> {
  type: ActionTypes.ADD_FAVORITE_STOCK;
  readonly stockSymbol: string;
}

/**
 * Remove stock from Favorites action
 */
export interface RemoveFavoriteStockAction extends Action<ActionTypes.REMOVE_FAVORITE_STOCK> {
  type: ActionTypes.REMOVE_FAVORITE_STOCK;
  readonly stockSymbol: string;
}

/**
 * Clear all stocks from Favorites action
 */
export interface ClearFavoriteStocksAction extends Action<ActionTypes.CLEAR_FAVORITE_STOCKS> {
  type: ActionTypes.CLEAR_FAVORITE_STOCKS;
}

/**
 * Union of actions for AddFavoriteStock, RemoveFavoriteStock, ClearFavoriteStocks
 */
export type FavoritesActions =
  | AddFavoriteStockAction
  | RemoveFavoriteStockAction
  | ClearFavoriteStocksAction;
