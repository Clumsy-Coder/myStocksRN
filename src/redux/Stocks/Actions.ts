/* eslint-disable max-len */
import { ActionCreator } from 'redux';

import { ActionTypes, DataDomain, Actions } from 'src/redux/Stocks/Types';

// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
//                                                                                                //
//                                  STOCK QUOTE ACTIONS CREATORS                                  //
//                                                                                                //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //

/**
 * Fetch stock quote action creator.
 * ONLY to be called by the front end.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 */
export const fetchStockQuote: ActionCreator<Actions.Quote.FetchAction> = (
  stockSymbol: string,
): Actions.Quote.FetchAction => ({
  type: ActionTypes.FETCH_STOCK_QUOTE,
  stockSymbol,
});

/**
 * Fetch stock quote pending action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 */
export const fetchStockQuotePending: ActionCreator<Actions.Quote.FetchPendingAction> = (
  stockSymbol: string,
): Actions.Quote.FetchPendingAction => ({
  type: ActionTypes.FETCH_STOCK_QUOTE_PENDING,
  stockSymbol,
});

/**
 * Fetch stock quote fulfilled action creator.
 * ONLY to be called by redux saga.
 * Only used internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param payload - data fetched
 */
export const fetchStockQuoteFulfilled: ActionCreator<Actions.Quote.FetchFulfilledAction> = (
  stockSymbol: string,
  payload: DataDomain.StockQuote,
): Actions.Quote.FetchFulfilledAction => ({
  type: ActionTypes.FETCH_STOCK_QUOTE_FULFILLED,
  stockSymbol,
  payload,
});

/**
 * Fetch stock quote rejected action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param error - error message
 */
export const fetchStockQuoteRejected: ActionCreator<Actions.Quote.FetchRejectedAction> = (
  stockSymbol: string,
  error: Error,
): Actions.Quote.FetchRejectedAction => ({
  type: ActionTypes.FETCH_STOCK_QUOTE_REJECTED,
  stockSymbol,
  error,
});

// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
//                                                                                                //
//                             STOCK DAILY ADJUSTED ACTIONS CREATORS                              //
//                                                                                                //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //

/**
 * Fetch stock Daily Adjusted action creator.
 * ONLY to be called by the front end.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param outputsize - 'compact' or 'full'
 */
export const fetchStockDailyAdj: ActionCreator<Actions.DailyAdjusted.FetchAction> = (
  stockSymbol: string,
  outputsize: 'compact' | 'full',
): Actions.DailyAdjusted.FetchAction => ({
  type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED,
  stockSymbol,
  outputsize,
});

/**
 * Fetch stock Daily Adjusted pending action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 */
export const fetchStockDailyAdjPending: ActionCreator<Actions.DailyAdjusted.FetchPendingAction> = (
  stockSymbol: string,
): Actions.DailyAdjusted.FetchPendingAction => ({
  type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED_PENDING,
  stockSymbol,
});

/**
 * Fetch Stock Daily Adjusted fulfilled action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param payload - data fetched
 */
export const fetchStockDailyAdjFulfilled: ActionCreator<Actions.DailyAdjusted.FetchFulfilledAction> = (
  stockSymbol: string,
  payload: DataDomain.StockDailyAdj,
): Actions.DailyAdjusted.FetchFulfilledAction => ({
  type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED_FULFILLED,
  stockSymbol,
  payload,
});

/**
 * Fetch Stock Daily Adjusted rejected action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param error - error message
 */
export const fetchStockDailyAdjRejected: ActionCreator<Actions.DailyAdjusted.FetchRejectedAction> = (
  stockSymbol: string,
  error: Error,
): Actions.DailyAdjusted.FetchRejectedAction => ({
  type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED_REJECTED,
  stockSymbol,
  error,
});

// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
//                                                                                                //
//                                 STOCK SEARCH ACTIONS CREATORS                                  //
//                                                                                                //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //

export const fetchStockSearch: ActionCreator<Actions.Search.FetchAction> = (
  keyword: string,
): Actions.Search.FetchAction => ({
  type: ActionTypes.SEARCH_KEYWORD,
  keyword,
});

export const fetchStockSearchPending: ActionCreator<Actions.Search.FetchPendingAction> = (
  keyword: string,
): Actions.Search.FetchPendingAction => ({
  type: ActionTypes.SEARCH_KEYWORD_PENDING,
  keyword,
});

export const fetchStockSearchFulfilled: ActionCreator<Actions.Search.FetchFulfilledAction> = (
  keyword: string,
  payload: DataDomain.StockSearch,
): Actions.Search.FetchFulfilledAction => ({
  type: ActionTypes.SEARCH_KEYWORD_FULFILLED,
  keyword,
  payload,
});

export const fetchStockSearchRejected: ActionCreator<Actions.Search.FetchRejectedAction> = (
  keyword: string,
  error: Error,
): Actions.Search.FetchRejectedAction => ({
  type: ActionTypes.SEARCH_KEYWORD_REJECTED,
  keyword,
  error,
});

// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
//                                                                                                //
//                                  STOCK BATCH ACTIONS CREATORS                                  //
//                                                                                                //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //

// /**
//  * Fetch multiple stock quote data in a single batch action creator.
//  * ONLY to be called by the front end.
//  * @param stockSymbols - An array of company stock symbol in uppercase. Ex: ['AAPL', 'SHOP]
//  */
// export const fetchStockQuoteBatch: ActionCreator<Actions.Batch.FetchQuoteBatchAction> = (): Actions.Batch.FetchQuoteBatchAction => ({
//   type: ActionTypes.FETCH_STOCK_QUOTE_BATCH,
// });

// /**
//  * Fetch multiple stock chart data in a single batch action creator.
//  * ONLY to be called by the front end.
//  * @param stockSymbols - An array of company stock symbol in uppercase. Ex: ['AAPL', 'SHOP]
//  * @param range - Date range for the stock chart
//  * @param sort - sort data by date in ascending or descending order
//  */
// export const fetchStockChartBatch: ActionCreator<Actions.Batch.FetchDailyAdjustedBatchAction> = (): Actions.Batch.FetchDailyAdjustedBatchAction => ({
//   type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED_BATCH,
// });
