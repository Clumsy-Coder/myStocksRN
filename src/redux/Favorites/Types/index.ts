import ActionTypes from '@redux/Favorites/Types/ActionTypes';
import * as Actions from '@redux/Favorites/Types/Actions';
import * as Reducer from '@redux/Favorites/Types/Reducer';

export { ActionTypes };
export { Actions };
export { Reducer };
