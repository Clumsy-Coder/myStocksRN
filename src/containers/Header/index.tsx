import React, { useState } from 'react';
import { StyleSheet } from 'react-native';
import { Dispatch } from 'redux';
import { connect } from 'react-redux';
import { Header, Icon, Input, Item } from 'native-base';

import { AppState, AppActions } from '@redux/index.reducers';
import { Actions } from '@redux/Stocks/Types';
import { selectSearchKeyword } from '@redux/Stocks/Selectors';
import { setSearchKeyword, clearSearchKeyword } from '@redux/Stocks/Actions';

interface SelectorProps {
  searchKeyword: string;
}

interface DispatchProps {
  setKeyword: (keyword: string) => Actions.Search.SetSearchKeywordAction;
  clearKeyword: () => Actions.Search.ClearSearchKeywordAction;
}

type Props = SelectorProps & DispatchProps;

const styles = StyleSheet.create({
  headerBorder: {
    paddingVertical: 30,
    backgroundColor: 'white',
  },
});

const SearchHeader: React.FC<Props> = (props: Props) => {
  const { setKeyword, clearKeyword, searchKeyword } = props;
  const [inFocus, setFocus] = useState(false);

  return (
    <Header searchBar style={styles.headerBorder}>
      <Item>
        <Icon name='search' type='MaterialIcons' />
        <Input
          placeholder='Search'
          onFocus={(): void => setFocus(true)}
          onBlur={(): void => {
            clearKeyword();
            setFocus(false);
          }}
          onChangeText={setKeyword}
          value={searchKeyword}
        />
        {inFocus ? <Icon name='clear' type='MaterialIcons' onPress={clearKeyword} /> : undefined}
      </Item>
    </Header>
  );
};

export const mapStateToProps = (state: AppState): SelectorProps => ({
  searchKeyword: selectSearchKeyword(state),
});

export const mapDispatchToProps = (dispatch: Dispatch<AppActions>): DispatchProps => ({
  setKeyword: (keyword: string): Actions.Search.SetSearchKeywordAction =>
    dispatch(setSearchKeyword(keyword)),
  clearKeyword: (): Actions.Search.ClearSearchKeywordAction => dispatch(clearSearchKeyword()),
});

export default connect(mapStateToProps, mapDispatchToProps)(SearchHeader);
