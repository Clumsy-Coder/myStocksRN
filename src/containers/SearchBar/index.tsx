import React from 'react';
import { StyleSheet } from 'react-native';
import { Dispatch } from 'redux';
import { connect } from 'react-redux';
import { Item } from 'native-base';
import AutoComplete from 'react-native-autocomplete-input';

import SearchResultItem from '@components/SearchResultItem';
import { AppState, AppActions } from '@redux/index.reducers';
import { Actions, DataDomain } from '@redux/Stocks/Types';
import { selectSearchKeyword, selectSymbolsMetadataData } from '@redux/Stocks/Selectors';
import { setSearchKeyword, fetchStockQuote } from '@redux/Stocks/Actions';
import { filterSearch } from '@share/Utilities';

interface SelectorProps {
  searchKeyword: string;
  symbolsMetadata: DataDomain.Symbols[];
}

interface DispatchProps {
  setKeyword: (keyword: string) => Actions.Search.SetSearchKeywordAction;
  fetchQuote: (symbol: string) => Actions.Quote.FetchAction;
}

type Props = SelectorProps & DispatchProps;

const styles = StyleSheet.create({
  header: {
    paddingVertical: 10,
    paddingHorizontal: 20,
    borderWidth: 0,
  },
});

const SearchHeader: React.FC<Props> = (props: Props) => {
  const { setKeyword, fetchQuote, searchKeyword, symbolsMetadata } = props;

  return (
    <Item>
      <AutoComplete
        placeholder='Search'
        data={filterSearch(searchKeyword, symbolsMetadata)}
        onChangeText={setKeyword}
        renderItem={(obj: { item: DataDomain.Symbols }) => (
          <SearchResultItem data={obj.item} onPress={() => fetchQuote(obj.item.symbol)} />
        )}
        inputContainerStyle={styles.header}
        value={searchKeyword}
        keyExtractor={(item) => `search-result-${item.symbol}`}
      />
    </Item>
  );
};

export const mapStateToProps = (state: AppState): SelectorProps => ({
  searchKeyword: selectSearchKeyword(state),
  symbolsMetadata: selectSymbolsMetadataData(state),
});

export const mapDispatchToProps = (dispatch: Dispatch<AppActions>): DispatchProps => ({
  setKeyword: (keyword: string) => dispatch(setSearchKeyword(keyword)),
  fetchQuote: (symbol: string) => dispatch(fetchStockQuote(symbol)),
});

export default connect(mapStateToProps, mapDispatchToProps)(SearchHeader);
