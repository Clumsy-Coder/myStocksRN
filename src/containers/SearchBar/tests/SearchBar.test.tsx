import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';

import { SearchBar, mapStateToProps, mapDispatchToProps } from 'src/containers/SearchBar';
import { AppState } from 'src/redux/index.reducers';
import { Actions } from 'src/redux/Stocks/Types';
import { fetchStockQuote, setSearchKeyword } from 'src/redux/Stocks/Actions';

import * as testdata from 'jest.testdata';

describe('<SearchBar />', () => {
  it('Render correctly', () => {
    const props = {
      searchKeyword: '',
      symbolsMetadata: [
        testdata.symbolsMetadata1,
        testdata.symbolsMetadata2,
        testdata.symbolsMetadata3,
      ],
      setKeyword: jest.fn(),
      fetchQuote: jest.fn(),
    };
    const wrapper = shallow(<SearchBar {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {
    it('All properties are present', () => {
      const rootState: AppState = {
        Stocks: {
          search: {
            keyword: '',
          },
          symbols: {},
          symbolsMetadata: {
            fetching: false,
            error: undefined,
            data: [testdata.symbolsMetadata1, testdata.symbolsMetadata2, testdata.symbolsMetadata3],
          },
        },
        Favorites: {
          symbols: [],
        },
      };

      const stateToProps = mapStateToProps(rootState);
      expect(stateToProps).toHaveProperty('searchKeyword');
      expect(stateToProps).toHaveProperty('symbolsMetadata');
      expect(stateToProps.searchKeyword).toEqual('');
      expect(stateToProps.symbolsMetadata).toEqual([
        testdata.symbolsMetadata1,
        testdata.symbolsMetadata2,
        testdata.symbolsMetadata3,
      ]);
    });
  });

  describe('mapDispatchToProps', () => {
    it('All dispatch properties are present', () => {
      const dispatch = jest.fn();

      interface DispatchProps {
        setKeyword: (keyword: string) => Actions.Search.SetSearchKeywordAction;
        fetchQuote: (symbol: string) => Actions.Quote.FetchAction;
      }

      const dispatchToProps: DispatchProps = mapDispatchToProps(dispatch);
      dispatchToProps.fetchQuote(testdata.stockSymbol1);
      dispatchToProps.setKeyword(testdata.stockSymbol1);

      expect(dispatch.mock.calls.length).toEqual(2);
      expect(dispatch.mock.calls[0][0]).toEqual(fetchStockQuote(testdata.stockSymbol1));
      expect(dispatch.mock.calls[1][0]).toEqual(setSearchKeyword(testdata.stockSymbol1));
    });
  });
});
