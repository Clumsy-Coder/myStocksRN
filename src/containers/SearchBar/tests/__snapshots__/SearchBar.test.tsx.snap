// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> Render correctly 1`] = `
<Styled(Item)>
  <Autocomplete
    data={Array []}
    flatListProps={Object {}}
    inputContainerStyle={
      Object {
        "borderWidth": 0,
        "paddingHorizontal": 20,
        "paddingVertical": 10,
      }
    }
    keyExtractor={[Function]}
    keyboardShouldPersistTaps="always"
    onChangeText={[MockFunction]}
    onStartShouldSetResponderCapture={[Function]}
    placeholder="Search"
    renderItem={[Function]}
    renderSeparator={null}
    renderTextInput={[Function]}
  />
</Styled(Item)>
`;
